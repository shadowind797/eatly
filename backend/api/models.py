from django.db import models
from django.contrib.auth.models import AbstractUser


class User(AbstractUser):
    username = models.CharField(max_length=120, unique=True)
    password = models.CharField(max_length=120)
    phone = models.CharField(max_length=120, blank=True)
    status = models.ForeignKey('Status', on_delete=models.PROTECT, default=5)
    is_banned = models.BooleanField(default=False)
    ban_reason = models.TextField(blank=True)
    passwd_change_link = models.CharField(max_length=120, blank=True)

    def __str__(self):
        return self.username


class Item(models.Model):
    title = models.CharField(max_length=120, unique=True)
    description = models.TextField()
    price = models.FloatField()
    photo = models.ImageField(upload_to='img')
    category = models.ForeignKey('Category', on_delete=models.SET_NULL, null=True, blank=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True, blank=True)
    rating = models.FloatField()

    def __str__(self):
        return self.title


class Restaurant(models.Model):
    name = models.CharField(max_length=120, unique=True)
    address = models.CharField(max_length=120)
    rating = models.FloatField()
    image = models.ImageField(upload_to='img/rests', null=True)
    category = models.ForeignKey('RestaurantCat', on_delete=models.SET_NULL, null=True, blank=True)

    def __str__(self):
        return self.name


class Address(models.Model):
    house_address = models.CharField(max_length=120)
    entrance = models.CharField(max_length=120)
    floor = models.CharField(max_length=120)
    flat = models.CharField(max_length=120)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)

    def __str__(self):
        return self.house_address


class Order(models.Model):
    user = models.ForeignKey('User', on_delete=models.SET_NULL, null=True)
    payment = models.ForeignKey('Payments', on_delete=models.SET_NULL, null=True)
    address = models.ForeignKey('Address', on_delete=models.SET_NULL, null=True)
    total = models.FloatField()
    status = models.ForeignKey('OrderStatus', on_delete=models.SET_NULL, null=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True)
    coupon = models.ForeignKey('Coupon', on_delete=models.SET_NULL, null=True)
    comment = models.TextField(null=True, blank=True)

    def __str__(self):
        return self.user


class OrderStatus(models.Model):
    name = models.CharField(max_length=120, unique=True)


class Coupon(models.Model):
    title = models.CharField(max_length=120, unique=True)
    valid_from = models.DateField(auto_now_add=True)
    valid_to = models.DateField()
    is_valid = models.BooleanField(default=True)
    value = models.FloatField()
    times_activated = models.IntegerField(default=0)
    enabled_times = models.IntegerField()
    category = models.ForeignKey('CouponCategory', on_delete=models.CASCADE)

    def __str__(self):
        return self.title


class CouponCategory(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Payments(models.Model):
    number = models.CharField(max_length=120, unique=True)
    date_to = models.CharField(max_length=100)
    cvv = models.IntegerField()
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)
    name = models.CharField(max_length=120, default="")

    def __str__(self):
        return self.number


class CartItem(models.Model):
    item = models.ForeignKey('Item', on_delete=models.CASCADE)
    quantity = models.IntegerField(default=1)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)

    def __str__(self):
        return self.item.title


class Reviews(models.Model):
    author = models.ForeignKey('User', on_delete=models.CASCADE)
    time = models.CharField(max_length=120)
    text = models.TextField()
    rating = models.FloatField()
    created_at = models.DateTimeField(auto_now_add=True)

    def __str__(self):
        return self.author.username


class RestaurantCat(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Category(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Status(models.Model):
    name = models.CharField(max_length=120)

    def __str__(self):
        return self.name
