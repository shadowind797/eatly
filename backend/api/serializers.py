from .models import *
from rest_framework import serializers


class UserSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ['id', 'username', 'email', 'password', "phone", "status", "is_banned", "ban_reason", "first_name"]
        extra_kwargs = {"password": {"write_only": True}, "status": {"read_only": True}}

    def create(self, validated_data):
        user = User.objects.create_user(**validated_data)
        return user


class AddressSerializer(serializers.ModelSerializer):
    class Meta:
        model = Address
        fields = ["id", "house_address", "entrance", "floor", "flat"]

    def create(self, validated_data):
        address = Address.objects.create_address(**validated_data)
        return address


class PaymentSerializer(serializers.ModelSerializer):
    class Meta:
        model = Payments
        fields = "__all__"
        extra_kwargs = {"cvv": {"write_only": True}, "date_to": {"read_only": True}}

    def create(self, validated_data):
        payment = Payments.objects.create_payments(**validated_data)
        return payment


class OrderSerializer(serializers.ModelSerializer):
    class Meta:
        model = Order
        fields = ["id", "status", "total", "payment", "address"]

    def create(self, validated_data):
        order = Order.objects.create_order(**validated_data)
        return order


class ItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = Item
        fields = ['id', "title", "description", "price", "photo", "category", "rating"]


class CartItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = CartItem
        fields = ['id', "item", "quantity"]


class CategorySerializer(serializers.ModelSerializer):
    class Meta:
        model = Category
        fields = ['id', "name"]


class RestaurantSerializer(serializers.ModelSerializer):
    class Meta:
        model = Restaurant
        fields = ['id', "name", "address", "rating", "category_id", "image"]


class RestaurantCatSerializer(serializers.ModelSerializer):
    class Meta:
        model = RestaurantCat
        fields = ['id', "name"]
