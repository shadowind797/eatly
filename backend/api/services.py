from django.conf import settings
from django.shortcuts import redirect
from django.core.exceptions import ValidationError
from urllib.parse import urlencode
from typing import Dict, Any
import requests
from .models import User
import jwt

GOOGLE_ACCESS_TOKEN_OBTAIN_URL = 'https://oauth2.googleapis.com/token'
GOOGLE_USER_INFO_URL = 'https://www.googleapis.com/oauth2/v3/userinfo'

GITHUB_ACCESS_TOKEN_OBTAIN_URL = 'https://github.com/login/oauth/access_token'
GITHUB_USER_INFO_URL = 'https://api.github.com/user'

LOGIN_URL = f'{settings.BASE_APP_URL}/login'


def google_get_access_token(code: str, redirect_uri: str) -> str:
    data = {
        'code': code,
        'client_id': settings.GOOGLE_OAUTH2_CLIENT_ID,
        'client_secret': settings.GOOGLE_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri,
        'grant_type': 'authorization_code'
    }

    response = requests.post(GOOGLE_ACCESS_TOKEN_OBTAIN_URL, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Google.')

    access_token = response.json()['access_token']

    return access_token


def google_get_user_info(access_token: str) -> Dict[str, Any]:
    response = requests.get(
        GOOGLE_USER_INFO_URL,
        params={'access_token': access_token}
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Google.')

    return response.json()


def github_get_access_token(code: str, redirect_uri: str) -> str:
    headers = {
        "Accept": "application/json"
    }
    data = {
        'code': code,
        'client_id': settings.GITHUB_OAUTH2_CLIENT_ID,
        'client_secret': settings.GITHUB_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri
    }

    response = requests.post(GITHUB_ACCESS_TOKEN_OBTAIN_URL, headers=headers, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Github.')

    access_token = response.json()['access_token']

    return access_token


def github_get_user_info(access_token: str) -> Dict[str, Any]:
    headers = {
        'Authorization': f'Bearer {access_token}'
    }
    response = requests.get(
        GITHUB_USER_INFO_URL,
        headers=headers
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Github.')

    return response.json()


def get_user_data(validated_data, platform):
    domain = settings.BASE_API_URL
    redirect_uri = f'{domain}/api/login/{platform}/'

    code = validated_data.get('code')
    error = validated_data.get('error')
    state = validated_data.get('state')

    if error or not code:
        params = urlencode({'error': error})
        return redirect(f'{LOGIN_URL}?{params}')

    if platform == "google":
        access_token = google_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = google_get_user_info(access_token=access_token)

        user, created = User.objects.get_or_create(
            email=user_data['email'],
        )

        if not created:
            user.username = user_data['name']
            user.save()

        profile_data = {
            'email': user_data['email'],
        }
        return profile_data
    elif platform == "github":
        access_token = github_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = github_get_user_info(access_token=access_token)

        print(user_data)

        User.objects.get_or_create(
            username=user_data['login'],
        )

        profile_data = {
            'username': user_data['login'],
        }
        return profile_data
