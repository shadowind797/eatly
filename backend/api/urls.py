from django.urls import path, include

from . import views

urlpatterns = [
    path('access/', views.GetAccess.as_view(), name='access'),
    path('user/', views.GetUser.as_view(), name='user'),
    path('user/change/', views.GetUser.as_view(), name='user-change'),

    path("login/google/", views.GoogleLoginApi.as_view(), name='google-login'),
    path("login/github/", views.GithubLoginApi.as_view(), name='github-login'),

    path('items/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/top/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/category/', views.CategoriesList.as_view(), name='item-category-list'),

    path('items/cart/', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/add', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/delete', views.DeleteCartItem.as_view(), name='item-cart-list'),
    path('items/cart/check', views.CheckInCart.as_view(), name='item-cart-list'),

    path("items/delete/<int:pk>/", views.ItemDelete.as_view(), name='item-delete'),

    path("restaurants/", views.RestaurantListCreate.as_view(), name='restaurant-list'),
    path("restaurants/categories/", views.RestaurantCategories.as_view(), name='restaurant-cats-list'),

    path("address/", views.AddressList.as_view(), name='address-list'),
    path("address/add/", views.AddressList.as_view(), name='restaurant-cats-list'),

    path("order/", views.OrderView.as_view(), name='order'),
    path("order/add/", views.OrderView.as_view(), name='order-create'),
    path("order/cancel/", views.CancelOrder.as_view(), name='order-cancel'),

    path("payment/", views.PaymentView.as_view(), name='payment'),
    path("payment/add/", views.PaymentView.as_view(), name='payment-create'),

    path("coupon/", views.CouponListCreate.as_view(), name='coupon-list'),
    path("coupon/add/", views.CouponListCreate.as_view(), name='coupon-create'),

    path("items/search/", views.SearchView.as_view(), name='search'),
    path("items/search/filters/", views.FilterView.as_view(), name='filter'),

    path("profile/orders", views.GetProfile.as_view(), name='all-user-orders'),
]
