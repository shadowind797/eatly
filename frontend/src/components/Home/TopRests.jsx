import {useEffect, useState} from "react";
import RestaurantCard from "../RestaurantCard.jsx";
import api from "../../api.js";

function TopRests() {
    const [rests, setRests] = useState([]);

    useEffect(() => {
        getRests()
    }, []);

    const getRests = () => {
        api
            .get("api/restaurants/")
            .then((res) => res.data)
            .then((data) => setRests(data))
            .catch((err) => alert(err));
    }

    const topRests = rests.sort((a, b) => (b.rating - a.rating))
    const slicedRests = topRests.slice(0, 3);

    return (
        <div id="top-rests" className="container">
            <h2>Our Top <span>Restaurants</span></h2>
            <div id="tops">
                {slicedRests.map((rest) => <RestaurantCard rest={rest} />)}
            </div>
        </div>
    )
}

export default TopRests