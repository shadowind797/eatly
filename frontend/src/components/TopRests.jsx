import {useEffect, useState} from "react";
import Restaurant from "./Restaurant.jsx";
import api from "../api.js";
import header_load from "../assets/header-loading.gif"
import ViewAll from "../assets/view-all.svg"

function TopRests() {
    const [rests, setRests] = useState([]);
    const [cats, setCats] = useState([]);
    const [loading, setLoading] = useState(false)

    useEffect(() => {
        getRests()
    }, []);

    const getRests = () => {
        setLoading(true)
        api
            .get("api/restaurants/", {params: {method: "top"}})
            .then((res) => res.data)
            .then((data) => {
                setRests(data.items)
                setCats(data.cats)
                setLoading(false)
            })
            .catch((err) => {
            });
    }

    const getCategory = (categoryId) => {
        const cat = cats.find((cat) => cat.id === categoryId) || {name: ""}
        return cat.name;
    }

    if (loading) {
        return (
            <div id="top-rests" className="container">
                <h2>Our Top <span>Restaurants</span></h2>
                <img src={header_load} style={{width: "300px"}} alt=""/>
            </div>
        )
    } else {
        return (
            <div id="top-rests" className="container">
                <h2>Our Top <span>Restaurants</span></h2>
                <div id="tops">
                    {rests.map((rest) => <Restaurant rest={rest}
                                                     key={rest.id}
                                                     category={getCategory(rest.category_id)}/>)}
                </div>
                <div className="view-all">
                    <a href="#">
                        <p>View All</p>
                        <img src={ViewAll} alt=""/>
                    </a>
                </div>
            </div>
        )
    }
}

export default TopRests