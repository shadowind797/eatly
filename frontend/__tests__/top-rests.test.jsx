import React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react";
import api from "../src/api.js";
import { MemoryRouter as Router } from "react-router-dom";
import TopRests from "../src/components/TopRests.jsx";

jest.mock("../src/api");

describe("Top-Restaurants component", () => {
  beforeEach(() => {
    api.get.mockResolvedValue({
      data: {
        items: [
          {
            id: 2,
            name: "The Burger King",
            address: "Podshobog ko'chasi 23, Тоshkent, Toshkent, Uzbekistan",
            rating: 4.9,
            category_id: 2,
            image: "/media/img/rests/mcdonalds.jpg",
            items_count: 2,
          },
          {
            id: 5,
            name: "Maurice",
            address: "Bobur ko'chasi 34, Tashkent, Uzbekistan",
            rating: 4.9,
            category_id: 3,
            image: "/media/img/rests/mcdonalds.jpg",
            items_count: 3,
          },
          {
            id: 1,
            name: "The Chicken King",
            address: "59, Buyuk Ipak Yo'li 60, Тоshkent, Toshkent, Uzbekistan",
            rating: 4.8,
            category_id: 1,
            image: "/media/img/rests/burgerking.jpg",
            items_count: 3,
          },
        ],
        cats: [
          {
            id: 1,
            name: "Healthy",
          },
          {
            id: 3,
            name: "Supreme",
          },
          {
            id: 2,
            name: "Trending",
          },
        ],
      },
    });
  });

  it("renders without crashing", async () => {
    await act(async () => {
      render(
        <Router>
          <TopRests />
        </Router>
      );
    });
  });

  it("matches the snapshot", async () => {
    let asFragment;
    await act(async () => {
      const rendered = render(
        <Router>
          <TopRests />
        </Router>
      );
      asFragment = rendered.asFragment;
    });

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
