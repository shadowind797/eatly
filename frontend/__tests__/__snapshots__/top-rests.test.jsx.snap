// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top-Restaurants component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    id="top-rests"
  >
    <h2>
      Our Top 
      <span>
        Restaurants
      </span>
    </h2>
    <div
      id="tops"
    >
      <div
        class="rest-card"
      >
        <img
          alt=""
          class="rest-img"
          src="undefined/media/img/rests/mcdonalds.jpg"
        />
        <div
          class="info"
        >
          <div
            class="cat"
            style="background-color: rgb(247, 197, 186); color: rgb(251, 71, 29);"
          >
            Trending
          </div>
          <div
            class="text"
          >
            <h4>
              The Burger King
            </h4>
            <div
              class="widgets"
            >
              <div
                class="time-rating"
              >
                <p>
                  2 dishes
                </p>
                <div>
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                  <p>
                    4.9
                  </p>
                </div>
              </div>
              <button>
                <img
                  alt=""
                  src="test-file-stub"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rest-card"
      >
        <img
          alt=""
          class="rest-img"
          src="undefined/media/img/rests/mcdonalds.jpg"
        />
        <div
          class="info"
        >
          <div
            class="cat"
            style="background-color: rgb(247, 237, 208); color: rgb(218, 163, 26);"
          >
            Supreme
          </div>
          <div
            class="text"
          >
            <h4>
              Maurice
            </h4>
            <div
              class="widgets"
            >
              <div
                class="time-rating"
              >
                <p>
                  3 dishes
                </p>
                <div>
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                  <p>
                    4.9
                  </p>
                </div>
              </div>
              <button>
                <img
                  alt=""
                  src="test-file-stub"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rest-card"
      >
        <img
          alt=""
          class="rest-img"
          src="undefined/media/img/rests/burgerking.jpg"
        />
        <div
          class="info"
        >
          <div
            class="cat"
            style="background-color: rgba(44, 196, 105, 0.45); color: rgb(48, 157, 91);"
          >
            Healthy
          </div>
          <div
            class="text"
          >
            <h4>
              The Chicken King
            </h4>
            <div
              class="widgets"
            >
              <div
                class="time-rating"
              >
                <p>
                  3 dishes
                </p>
                <div>
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                  <p>
                    4.8
                  </p>
                </div>
              </div>
              <button>
                <img
                  alt=""
                  src="test-file-stub"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="view-all"
    >
      <a
        href="/menu"
      >
        <p>
          View All
        </p>
        <img
          alt=""
          src="test-file-stub"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
